README.md
# NAME

WebService::ForecastIO - Perl client for api.forecast.io

# VERSION

version 0.02

# SYNOPSIS

    use 5.014;

    my $forecast = WebService::ForecastIO->new(
            api_key => 'secret',
            units => 'si', # metric units
            exclude => 'hourly,minutely,currently,flags', # get 7 day forecast only
    );

    #                             latitude          longitude
    my $json = $forecast->request(29.7492738082192, -95.4709680410959);

    # Human readable summary of daily data
    say $json->{daily}->{summary};

# OVERVIEW

This is a Perl client for [forecast.io API](https://developer.forecast.io). Forecast.io 
applies "big data" analysis techniques to publicly available weather data including 
radar image analysis.  One of the things it attempts to predict is _when_ certain
weather events like rain will start, and the duration of those events.

This library requires an API key which can be obtained for free from the 
[developer web site](https://developer.forecast.io). The first 1,000 calls
per day are allowed without charge.  (More calls can be made if payment arrangements
are made.)

See the [API docs](https://developer.forecast.io/docs/v2) for full details about
what data is provided and what granularity data sets are offered.

**NOTE**: Errors are fatal. Please use something like [Try::Tiny](https://metacpan.org/pod/Try::Tiny) if you
want to handle errors some other way.

# ATTRIBUTES

## units

This attribute overrides the default units of `us`. Available units are:

- `si`

    Metric units (windspeed in meters/second)

- `ca`

    Metric units, except windspeed in km/h

- `uk`

    Metric units, except windspeed in mi/h

- `auto`

    Unit selection based on request IP geolocation

## exclude

A comma delimited list of data sets to exclude. Choices are:

`minutely`, `currently`, `hourly`, `daily`, `alerts`, `flags`

See the API docs for information about what these data blocks represent.

Spaces should be omitted between exclusions.

## api\_key

Holds the API key for the forecast.io service.

# METHODS

## request

This method takes two or three parameters.

The first two parameters are `latitude` and `longitude` expressed
as floats. The optional third parameter is a `time` value. The time
value can be expressed as either epoch seconds or as an iso8601
format time. 

Data sets up to 60 years old are available. (Such times should
be expressed as iso8601 format times since epoch seconds start
at midnight, January 1, 1970.)

## to\_timepiece

A convenience method that takes epoch seconds and returns a
new [Time::Piece](https://metacpan.org/pod/Time::Piece) object.  You can stringify this object to
iso8601 format by calling the `datetime()` method on it.

## parse\_datetime

A convenience method that takes an iso8601 formatted string
and returns a new [Time::Piece](https://metacpan.org/pod/Time::Piece) object. This object can output
epoch seconds by calling the `epoch()` method on it.

# SEE ALSO

- [WebService::ForecastIO::Request](https://metacpan.org/pod/WebService::ForecastIO::Request)
- [API docs](https://developer.forecast.io/docs/v2)

# AUTHOR

Mark Allen <mrallen1@yahoo.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Mark Allen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Change history for WebService-ForecastIO

{{$NEXT}}

0.02        2016-11-21
            API URL updated

0.01        2013-05-08
            Initial release


lib/WebService/DarkSky.pm
use 5.014;

package WebService::DarkSky;

use Moo;
with 'WebService::DarkSky::Request';

use Time::Piece;

# ABSTRACT: Perl client for api.DarkSky.net

=head1 SYNOPSIS

    use 5.014;

    my $forecast = WebService::DarkSky->new(
            api_key => 'secret',
            units => 'si', # metric units
            exclude => 'hourly,minutely,currently,flags', # get 7 day forecast only
    );

    #                             latitude          longitude
    my $json = $forecast->request(29.7492738082192, -95.4709680410959);

    # Human readable summary of daily data
    say $json->{daily}->{summary};

=head1 OVERVIEW

This is a Perl client for [darksky.net API](https://darksky.net/dev/). DarkSky formally Forecast.io 
applies "big data" analysis techniques to publicly available weather data including 
radar image analysis.  One of the things it attempts to predict is _when_ certain
weather events like rain will start, and the duration of those events.

This library requires an API key which can be obtained for free from the 
[developer web site](https://darksky.net/dev/docs) after sign up. The first 1,000 calls
per day are allowed without charge.  (More calls can be made if payment arrangements
are made.)

See the [API docs](https://darksky.net/dev/docs/response) for full details about
what data is provided and what granularity data sets are offered.

B<NOTE>: Errors are fatal. Please use something like L<Try::Tiny> if you
want to handle errors some other way.

=attr units

This attribute overrides the default units of C<us>. Available units are:

=over

=item * C<si>

Metric units (windspeed in meters/second)

=item * C<ca>

Metric units, except windspeed in km/h

=item * C<uk>

Metric units, except windspeed in mi/h

=item * C<auto>

Unit selection based on request IP geolocation

=back

=cut

has 'units' => (
    is => 'ro',
);

=attr exclude

A comma delimited list of data sets to exclude. Choices are:

C<minutely>, C<currently>, C<hourly>, C<daily>, C<alerts>, C<flags>

See the API docs for information about what these data blocks represent.

Spaces should be omitted between exclusions.

=cut

has 'exclude' => (
    is => 'ro',
);

=attr api_key

Holds the API key for the forecast.io service.

=method request

This method takes two or three parameters.

The first two parameters are C<latitude> and C<longitude> expressed
as floats. The optional third parameter is a C<time> value. The time
value can be expressed as either epoch seconds or as an iso8601
format time. 

Data sets up to 60 years old are available. (Such times should
be expressed as iso8601 format times since epoch seconds start
at midnight, January 1, 1970.)

=method to_timepiece

A convenience method that takes epoch seconds and returns a
new L<Time::Piece> object.  You can stringify this object to
iso8601 format by calling the C<datetime()> method on it.

=cut

sub to_timepiece {
    my ($self, $epoch_secs) = @_;

    return Time::Piece->new($epoch_secs);
}

=method parse_datetime

A convenience method that takes an iso8601 formatted string
and returns a new L<Time::Piece> object. This object can output
epoch seconds by calling the C<epoch()> method on it.

=cut

sub parse_datetime {
    my ($self, $iso8601) = @_;

    return Time::Piece->strptime($iso8601, "%Y-%m-%dT%H:%M:%S");
}

=head1 SEE ALSO

=over

=item * L<WebService::DarkSky::Request>

=item * L<API docs|https://darksky.net/dev/>

=back

=cut

1;


lib/WebService/ForecastIO/Request.pm
use 5.014;

package WebService::ForecastIO::Request;

use Moo::Role;
use HTTP::Tiny;
use JSON;

# ABSTRACT: Request role for WebService::ForecaseIO

=head1 OVERVIEW

This is a role which implements requests to the L<forecast.io> API.

=attr base_url

The base url to connect to the web service. Defaults to L<https://api.forecast.io/forecast>

=cut

has 'base_url' => (
    is => 'ro',
    default => sub { "https://api.darksky.net/forecast" },
);

has 'api_key' => (
    is => 'ro',
    required => 1
);

=attr ua

The user agent for the role. Uses L<HTTP::Tiny>.

=cut

has 'ua' => (
    is => 'ro',
    default => sub {
        HTTP::Tiny->new(
            agent => "WebService::ForecastIO/$WebService::ForecastIO::VERSION ",
            SSL_options => {
                SSL_hostname => "",
                SSL_verify_mode => 0
            },
        );
    },
    lazy => 1,
);

=attr decoder

The library to deserialize JSON responses. Uses L<JSON>.

=cut

has 'decoder' => (
    is => 'ro',
    default => sub {
        JSON->new();
    },
    lazy => 1,
);

sub request {
    my $self = shift;

    my $url = $self->base_url . "/" . $self->api_key . "/" . (join ",", @_);

    my $qp = join "&", (map {; "$_=" . $self->$_() } 
                        grep {; defined $self->$_() } qw(exclude units));

    if ( $qp ) {
        $url .= "?$qp";
    }

    my $response = $self->ua->get($url);

    if ( $response->{success} ) {
        $self->decoder->decode($response->{content});
    }
    else {
        die "Request to $url returned $response->{status}: $response->{content}\n";
    }
}

1;


lib/WebService/ForecastIO.pm
use 5.014;

package WebService::ForecastIO;

use Moo;
with 'WebService::ForecastIO::Request';

use Time::Piece;

# ABSTRACT: Perl client for api.forecast.io

=head1 SYNOPSIS

    use 5.014;

    my $forecast = WebService::ForecastIO->new(
            api_key => 'secret',
            units => 'si', # metric units
            exclude => 'hourly,minutely,currently,flags', # get 7 day forecast only
    );

    #                             latitude          longitude
    my $json = $forecast->request(29.7492738082192, -95.4709680410959);

    # Human readable summary of daily data
    say $json->{daily}->{summary};

=head1 OVERVIEW

This is a Perl client for L<forecast.io API|https://developer.forecast.io>. Forecast.io 
applies "big data" analysis techniques to publicly available weather data including 
radar image analysis.  One of the things it attempts to predict is I<when> certain
weather events like rain will start, and the duration of those events.

This library requires an API key which can be obtained for free from the 
L<developer web site|https://developer.forecast.io>. The first 1,000 calls
per day are allowed without charge.  (More calls can be made if payment arrangements
are made.)

See the L<API docs|https://developer.forecast.io/docs/v2> for full details about
what data is provided and what granularity data sets are offered.

B<NOTE>: Errors are fatal. Please use something like L<Try::Tiny> if you
want to handle errors some other way.

=attr units

This attribute overrides the default units of C<us>. Available units are:

=over

=item * C<si>

Metric units (windspeed in meters/second)

=item * C<ca>

Metric units, except windspeed in km/h

=item * C<uk>

Metric units, except windspeed in mi/h

=item * C<auto>

Unit selection based on request IP geolocation

=back

=cut

has 'units' => (
    is => 'ro',
);

=attr exclude

A comma delimited list of data sets to exclude. Choices are:

C<minutely>, C<currently>, C<hourly>, C<daily>, C<alerts>, C<flags>

See the API docs for information about what these data blocks represent.

Spaces should be omitted between exclusions.

=cut

has 'exclude' => (
    is => 'ro',
);

=attr api_key

Holds the API key for the forecast.io service.

=method request

This method takes two or three parameters.

The first two parameters are C<latitude> and C<longitude> expressed
as floats. The optional third parameter is a C<time> value. The time
value can be expressed as either epoch seconds or as an iso8601
format time. 

Data sets up to 60 years old are available. (Such times should
be expressed as iso8601 format times since epoch seconds start
at midnight, January 1, 1970.)

=method to_timepiece

A convenience method that takes epoch seconds and returns a
new L<Time::Piece> object.  You can stringify this object to
iso8601 format by calling the C<datetime()> method on it.

=cut

sub to_timepiece {
    my ($self, $epoch_secs) = @_;

    return Time::Piece->new($epoch_secs);
}

=method parse_datetime

A convenience method that takes an iso8601 formatted string
and returns a new L<Time::Piece> object. This object can output
epoch seconds by calling the C<epoch()> method on it.

=cut

sub parse_datetime {
    my ($self, $iso8601) = @_;

    return Time::Piece->strptime($iso8601, "%Y-%m-%dT%H:%M:%S");
}

=head1 SEE ALSO

=over

=item * L<WebService::ForecastIO::Request>

=item * L<API docs|https://developer.forecast.io/docs/v2>

=back

=cut

1;
